### ISSUE TITLE
Go to bottom isn't working properly

### ISSUE THREAD

**Issue by Vamshi99** [zulip/zulip-mobile#1974]
When I click `go to bottom` button, it just stops at last loaded message, if there are more messages to load it stops there and after a second, new messages will be there, but it doesn't scroll down.

**Comment by timabbott** [comment]
Yeah, that's definitely wrong.  In the webapp implementation we do something different since we always fetch up to current.  What we should do here is check if we have all messages in this view loaded (we should be able to know that state), and if not (1) send a request to the server to mark all messages in the view as read and then (2) re-narrow, fetching from server if required.

**Comment by borisyankov** [comment]
We can definitely do what Slack is doing, which is to go to bottom, load messages, go to bottom, load messages etc.

**Comment by timabbott** [comment]
Just to understand, the Slack thing sorta does it in a loop?  That seems not super great.

**Comment by borisyankov** [comment]
It is more obvious what is happening though, as visually it is scrolling. 
If we just replace the messages it will be visually confusing.

Somewhat related is that we have no good way to determine how many more messages might be beyond. We could have different behavior if there are a few hundred vs many more.

**Comment by borisyankov** [comment]
I removed the 'bug' label as currently, this functionality is working as intended.
We can rethink the behavior though.

**Comment by gnprice** [comment]
I got a report of this from a user who was frustrated that in a stream with like 4000 unread messages of history, there wasn't an obvious good way to skip to the end. (FWIW for anyone reading this with the same frustration: a workaround is to hit the "mark all as read" button, then navigate back out of the conversation/stream you're looking at, then back into it.)

See also the report at #2759 from @probabilistic back in 2018:
> Regarding the iOS app. Is there any way to make it jump to the most recent message? I know there is the little arrow at the bottom, but that only jumps to the most recent message that is loaded -- which is a problem if you're like me and only use the app once every few weeks or so.

I agree with @Vamshi99 and @timabbott above and with @probabilistic: the current behavior feels broken. Even if you understand what's happening and why, and you just keep hitting the jump-to-end button again and again, if you're in a view with a lot of history then it could take effectively forever to actually get to the bottom -- so it's not serving the user's purpose. If we were to imitate the behavior that Slack reportedly had in 2018, and basically automate that loop for you, it would still take effectively forever and still not serve the purpose.

Instead, when you hit the jump-to-end button, it should jump all the way to the end: to the latest messages that exist in the narrow. If we don't have the latest messages (i.e. if we're not caught up to latest in that narrow), then this will involve fetching the latest messages, from the end.

When doing so, it should also mark all the messages in the narrow as read -- just if we're currently marking messages as read on scroll. (This is typically true, but false in search narrows and some others, and false if the user has set the "Do not mark messages read on scroll" setting.)

(Not re-adding the "bug" label, just because that label no longer means anything in our workflow.)

---

Getting the UI for this right may require some fiddling. I agree with Boris above that there's a pitfall here:

> It is more obvious what is happening though, as visually it is scrolling. If we just replace the messages it will be visually confusing.

Still, I think even if we get a crude and jumpy experience but one that does take you promptly to the end, that will be an improvement over having to tap the button over and over and over again.

Probably a good solution will look something like:
* Scroll to the end of what we have. (This gives the visual of scrolling, and causes a loading spinner to appear at the bottom if we aren't at the end.)
* Mark all as read, if we're marking things as read on scroll in this view.
* If we don't have the latest messages (i.e. the `caughtUp` value for this narrow has `newest: false`), then start a fetch for the latest messages.
* When those messages arrive:
  * Discard the existing messages if they don't overlap with the ones we just got. (We only store one ["interval of completeness"](https://github.com/zulip/zulip-mobile/blob/5dc4770252efbb2cf5689be892948b44ff3e84c3/docs/architecture/realtime.md) per narrow.) I believe `narrowsReducer.js` will already take care of this.
  * Scroll to the end of the messages.

To accomplish the "fetch, and when complete, scroll to end", I believe it's enough to `await` the result of the appropriate fetch-messages action. Something like this:
```js
  await dispatch(fetchMessagesInNarrow(narrow, LAST_MESSAGE_ANCHOR));
  scrollToEnd(); // pseudocode
```
